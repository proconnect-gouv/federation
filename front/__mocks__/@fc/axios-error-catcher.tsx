import React, { PropsWithChildren } from 'react';

export const AxiosErrorCatcherContext = React.createContext(expect.any(Object));

export const AxiosErrorCatcherProvider = jest.fn(
  ({ children, value }: { value: any } & PropsWithChildren) => (
    <AxiosErrorCatcherContext.Provider value={value}>{children}</AxiosErrorCatcherContext.Provider>
  ),
);
