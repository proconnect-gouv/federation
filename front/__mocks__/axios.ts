const actualModule = jest.requireActual('axios');

module.exports = {
  ...actualModule,
  get: jest.fn(() => Promise.resolve(null)),
  interceptors: {
    request: {
      eject: jest.fn(),
      use: jest.fn(),
    },
    response: {
      eject: jest.fn(),
      use: jest.fn(),
    },
  },
  post: jest.fn(() => Promise.resolve(null)),
  request: jest.fn(() => Promise.resolve(null)),
};

export {};
