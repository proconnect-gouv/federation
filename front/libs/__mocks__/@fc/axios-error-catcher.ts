import React from 'react';

export const AxiosErrorCatcherProvider = jest.fn();

export const AxiosErrorCatcherContext = React.createContext(expect.any(Object));
