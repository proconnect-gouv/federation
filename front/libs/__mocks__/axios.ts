const actualModule = jest.requireActual('axios');

module.exports = {
  _esModule: true,
  ...actualModule,
  get: jest.fn(() => Promise.resolve(null)),
  post: jest.fn(() => Promise.resolve(null)),
};

export {};
