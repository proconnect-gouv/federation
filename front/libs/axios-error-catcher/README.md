# axios-error-catcher

Provides a react context to manage error request/response interceptors to axios.

## Context example

```
  <AxiosErrorCatcherProvider>
    <AxiosErrorCatcherContext.Consumer>
      {({ hasError, errorCode }) => (
        <div>...</div>
      )}
    </AxiosErrorCatcherContext.Consumer>
  </AxiosErrorCatcherProvider>
```
