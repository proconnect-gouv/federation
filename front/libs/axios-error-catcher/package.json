{
  "name": "@fc/axios-error-catcher",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "axios": "^0.27.2",
    "react": "^18.2.0"
  },
  "main": "src/index.ts"
}
