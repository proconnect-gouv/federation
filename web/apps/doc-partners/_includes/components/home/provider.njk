{% set item = params.item %}
<div class="fr-col-4">
  <div class="fr-card fr-card--sm fr-enlarge-link fr-card--no-icon">
    <div class="fr-card__body">
      <div class="fr-card__content">
        <h3 class="fr-card__title">
          {{ item.title }}
        </h3>
        <div class="fr-card__start">
          <ul class="fr-tags-group">
            {% for tag in item.tags %}
              {% if tag !== 'referencesFcFi' and tag !== 'referencesFcFs' %}
                <li>
                  <p class="fr-tag fr-tag--sm">
                    {{ tag }}
                  </p>
                </li>
              {% endif %}
            {% endfor %}
          </ul>
        </div>
      </div>
    </div>
    <div class="fr-card__header">
      <div class="fr-card__img">
        <img class="fr-responsive-img" src="images/home/{{ item.img }}" alt="{{ item.title }}" />
      </div>
    </div>
  </div>
</div>