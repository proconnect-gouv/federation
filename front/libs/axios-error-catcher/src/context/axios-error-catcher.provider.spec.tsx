import { render } from '@testing-library/react';
import axios from 'axios';
import React from 'react';

import { AxiosErrorCatcherProvider } from './axios-error-catcher.provider';

jest.mock('./axios-error-catcher.context');

describe('AxiosErrorCatcherProvider', () => {
  // given
  const Provider = () => (
    <AxiosErrorCatcherProvider>
      <div data-testid="ChildComponent" />
    </AxiosErrorCatcherProvider>
  );

  beforeEach(() => {
    // given
    jest
      .spyOn(React, 'useState')
      .mockReturnValue([{ codeError: undefined, hasError: false, initialized: false }, jest.fn()]);
  });

  it('should return null, if not initialized', () => {
    // given
    jest.spyOn(React, 'useState').mockReturnValueOnce([{ initialized: false }, jest.fn()]);

    // when
    const { container } = render(<Provider />);

    // then
    expect(container).toMatchSnapshot();
    expect(container.firstChild).toBeNull();
  });

  it('should render the child component, if initialized', () => {
    // given
    jest.spyOn(React, 'useState').mockReturnValueOnce([{ initialized: true }, jest.fn()]);

    // when
    const { container, getByTestId } = render(<Provider />);
    const element = getByTestId('ChildComponent');

    // then
    expect(container).toMatchSnapshot();
    expect(element).toBeInTheDocument();
  });

  it('should initialize the state', () => {
    // given
    const useStateMock = jest.spyOn(React, 'useState');

    // when
    render(<Provider />);

    // then
    expect(useStateMock).toHaveBeenCalledOnce();
    expect(useStateMock).toHaveBeenCalledWith({
      codeError: undefined,
      hasError: false,
      initialized: false,
    });
  });

  it('should initialize axios interceptors on first render only', () => {
    // given
    const useEffectMock = jest.spyOn(React, 'useEffect');

    // when
    const { rerender } = render(<Provider />);
    // @NOTE
    // excessive multiple rerenders (6)
    rerender(<Provider />);
    rerender(<Provider />);
    rerender(<Provider />);
    rerender(<Provider />);
    rerender(<Provider />);

    // then
    expect(useEffectMock).toHaveBeenCalledTimes(6);
    expect(useEffectMock).toHaveBeenCalledWith(expect.any(Function), []);

    expect(axios.interceptors.request.use).toHaveBeenCalledOnce();
    expect(axios.interceptors.request.use).toHaveBeenCalledWith(expect.any(Function), undefined);
    expect(axios.interceptors.response.use).toHaveBeenCalledOnce();
    expect(axios.interceptors.response.use).toHaveBeenCalledWith(undefined, expect.any(Function));
  });

  it('should eject axios interceptors on unmount', () => {
    // given
    const interceptorNumber = Symbol(123) as unknown as number;
    jest.mocked(axios.interceptors.request.use).mockReturnValue(interceptorNumber);
    jest.mocked(axios.interceptors.response.use).mockReturnValue(interceptorNumber);

    // when
    const { unmount } = render(<Provider />);
    unmount();

    // then
    expect(axios.interceptors.request.eject).toHaveBeenCalledOnce();
    expect(axios.interceptors.request.eject).toHaveBeenCalledWith(interceptorNumber);
    expect(axios.interceptors.response.eject).toHaveBeenCalledOnce();
    expect(axios.interceptors.response.eject).toHaveBeenCalledWith(interceptorNumber);
  });

  it('should update the state on first render only', () => {
    // given
    const setStateMock = jest.fn();
    jest.spyOn(React, 'useState').mockReturnValueOnce([expect.any(Object), setStateMock]);

    // when
    const { rerender } = render(<Provider />);
    // @NOTE
    // excessive multiple rerenders
    rerender(<Provider />);
    rerender(<Provider />);
    rerender(<Provider />);
    rerender(<Provider />);
    rerender(<Provider />);

    // then
    expect(setStateMock).toHaveBeenCalledOnce();
    expect(setStateMock).toHaveBeenCalledWith(expect.any(Function));

    // when
    // Check the function passed to setState
    const callback = setStateMock.mock.calls[0][0];
    const result = callback({ codeError: undefined, hasError: false, initialized: false });

    // then
    expect(result).toStrictEqual({ codeError: undefined, hasError: false, initialized: true });
  });
});
