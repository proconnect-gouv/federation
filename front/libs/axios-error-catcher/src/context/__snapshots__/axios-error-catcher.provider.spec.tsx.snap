// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AxiosErrorCatcherProvider should render the child component, if initialized 1`] = `
<div>
  <div
    data-testid="ChildComponent"
  />
</div>
`;

exports[`AxiosErrorCatcherProvider should return null, if not initialized 1`] = `<div />`;
